### DISCUSSION TITLE
JWT 토큰이 탈취 당한경우, 어떻게 해야 할까?

### DISCUSSION THREAD

**Discussion by beatmeJY** [dailyge/dailyge-server#114] [Information]
## 🔗 연관 PR
 - [[DLG-48] 만료된 AccessToken는 RefreshToken으로 토큰을 재발급한다.](https://github.com/dailyge/dailyge-server/pull/33)
 - [[DLG-183] JWT를 탈취 당할 경우, User 인증정보를 무효화 한다.](https://github.com/dailyge/dailyge-server/pull/116)

<br/><br/>

## ❓ 작성 이유
만일 서버에서 사용자 JWT 토큰이 탈취 당하였다고 판단이 되었을 경우에 사용자 정보를 모두 무효화 시켜주어야 합니다.
따라서 사용자의 RefreshToken만을 삭제한다면 유효 기간이 남아있는 AccessToken을 어떻게 무력화 시킬지에 대한 방법을 찾아보기 위해 작성하게 되었습니다.

<br/><br/>

### 장점
 - 만일 AccessToken이 탈취 당하더라도, 금방 효력을 잃어 사용할 수 없게 됩니다.
 - 레디스에 저장하고 있는 RefreshToken만 삭제하면 1분 내로 해당 사용자의 인증이 무효화가 될 수 있습니다.
### 단점
 - 거의 매 사용자의 요청이 AccessToken을 재 갱신 하므로 서버의 부하와 성능에 좋지 않습니다.
 - 그렇다고 수명을 더 늘리기에는 그 시간만큼 AccessToken으로 악의적인 행동을 할 수 있습니다.
 - 또한 정상 사용자가 긴 글을 작성할 경우 토큰이 만료되어 새로고침이 된다면 사용자 경험이 좋지 않을 수 있습니다.

<br/>
제가 최초에 생각한 방법으로 AccessToke을 1분 정도로 짧게 설정하면 되지 않을까 하였지만,

AccessToken으로만 사용자를 무효화를 시키는 것에는 위와 같은 한계가 있기 때문에, 해당 방법은 적정 시간으로 설정하여 보조적인 수단으로만 사용하는 것이 좋을 것 같습니다.🤔

<br/><br/>

## 2. 사용자 로그아웃 시간을 기록해두기.
> 사용자의 마지막 로그아웃 시간을 기록하고, 로그아웃 시간보다 이전에 생성된 토큰이라면 토큰을 무효화하는 방법입니다.
여기서는 무효화한 시간을 로그아웃 시간으로 설정하면 됩니다.

### 장점
 - 무효화를 하기 전에 발급된 토큰을 효과적으로 무효화 시킬 수 있습니다.
### 단점
 - 모든 사용자의 로그아웃 시간을 저장소에 시간을 기록해야 합니다.
 - 모든 사용자 요청 시 DB 조회가 발생하므로 성능 상 좋지 않을 수 있습니다.
 - 다중 기기에서 사용 중일 경우 다른 기기에서도 로그아웃이 될 수 있습니다.

<br/>

모든 사용자가 로그아웃을 하는 시간을 RDB에 저장하고, 매 요청시 비교를 위해 RDB에 조회를 하는 것은 성능적으로 좋지 못합니다.❌

<br/><br/>

### 장점
 - RefreshToken과 같이 캐시를 사용하여 빠르게 조회가 가능합니다.
 - 실제 무효화 대상은 상대적으로 적으므로 적은량의 메모리만 사용합니다.
 - 블랙리스트도 최대 AccessToken의 수명시간 만큼만 유지되도록 설정하면 되기에 추후 삭제하거나 관리포인트도 줄어듭니다.
 - 사용자가 블랙리스트에 있을 경우 블랙리스트 생성 시간을 기준으로 이전에 발급된 토큰만 무효화 시킬 수 있습니다.
### 단점
 - 모든 로그인 된 사용자는 매번 자신이 로그아웃 블랙리스트에 있는지 조회가 필요합니다.

성능적으로나 관리포인트면에서 **레디스에 무효화 블랙리스트**를 저장하여 비교하는 것이 좋을 것 같습니다.😄

**Comment by devjun10** [comment]
토큰을 1분 미만으로 발행하는 것은 현실적으로 불가능하겠죠?

**Reply by beatmeJY** [reply]
맞습니다 ㅎㅎ 계속해서 생각하고 찾다보니 위와 같은 단점이 있어서 다른 방법을 더 모색해보았습니다!👍🏻
